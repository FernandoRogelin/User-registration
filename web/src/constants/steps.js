export const FIRST_STEP = 1
export const SECOND_STEP = 2
export const THIRD_STEP = 3
export const LAST_STEP = 4

export const stepFields = {
  1: ['email', 'personType'],
  2: {
    person: ['name', 'cpf', 'birthdate', 'phone'],
    company: ['companyName', 'cnpj', 'openingDate', 'companyPhone']
  },
  3: ['password']
}
