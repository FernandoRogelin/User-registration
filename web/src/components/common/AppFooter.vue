<script setup>
import { computed } from 'vue'

import { stepFields, FIRST_STEP, LAST_STEP } from '@/constants/steps'

import Button from '@/components/ui/Button.vue'

const props = defineProps({
  step: {
    type: Number,
    required: true
  },
  personType: {
    type: String,
    required: true
  },
  nextStep: {
    type: Function,
    required: true
  },
  prevStep: {
    type: Function,
    required: true
  },
  validateField: {
    type: Function,
    required: true
  }
})

const emit = defineEmits(['submit'])

const isLastStep = computed(() => props.step === LAST_STEP || false)

async function onNextStep() {
  if (isLastStep.value) return emit('submit')

  let fieldsToValidate = stepFields[props.step]

  if (typeof fieldsToValidate === 'object' && !Array.isArray(fieldsToValidate)) {
    fieldsToValidate = fieldsToValidate[props.personType]
  }

  const validationPromises = fieldsToValidate.map(field => props.validateField(field))

  const results = await Promise.all(validationPromises)

  const isStepValid = results.every(result => result.valid)

  if (isStepValid) props.nextStep()
}
</script>

<template>
  <footer class="components-common-app-footer">
    <Button type="button" v-if="step > FIRST_STEP" variant="secondary" @click="prevStep">
      Voltar
    </Button>

    <Button :type="isLastStep ? 'submit' : 'button'" @click="onNextStep">
      {{ isLastStep ? 'Cadastrar' : 'Continuar' }}
    </Button>
  </footer>
</template>

<style scoped lang="scss">
.components-common-app-footer {
  gap: 1rem;
  display: flex;
}
</style>
