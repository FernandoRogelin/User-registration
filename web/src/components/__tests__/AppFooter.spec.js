import { describe, expect, test, afterEach, vi } from "vitest"
import { render, screen, cleanup, fireEvent } from "@testing-library/vue"

import AppFooter from "@/components/common/AppFooter.vue"

import { FIRST_STEP, SECOND_STEP } from "@/constants/steps"

afterEach(() => cleanup())

const nextStep = vi.fn()
const prevStep = vi.fn()
const validateField = vi.fn()

const mockProps = {
  step: FIRST_STEP,
  personType: 'company',
  nextStep,
  prevStep,
  validateField
}

describe('<AppFooter /> common component', () => {
  test('should render component and match snapshot', () => {
    const { container } = render(AppFooter, { props: mockProps })

    expect(container.firstChild).toMatchSnapshot()
  })

  test('should render back button and call prevStep function', async () => {
    render(AppFooter, { props: { ...mockProps, step: SECOND_STEP } })

    await fireEvent.click(screen.getByRole('button', { name: /voltar/i }))

    expect(prevStep).toHaveBeenCalled()
  })

  test('should call nextStep function when click on button', async () => {
    validateField.mockResolvedValue({ valid: true })
    render(AppFooter, { props: mockProps })

    await fireEvent.click(screen.getByRole('button', { name: /continuar/i }))

    expect(nextStep).toHaveBeenCalled()
  })
})
