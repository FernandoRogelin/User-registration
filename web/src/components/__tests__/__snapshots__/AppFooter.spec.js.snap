// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AppFooter /> common component > should render component and match snapshot 1`] = `
<footer
  class="components-common-app-footer"
  data-v-add55efe=""
>
  <!--v-if-->
  <button
    class="components-ui-button components-ui-button--primary"
    data-v-add55efe=""
    data-v-e3216c45=""
    type="button"
  >
    
    Continuar
    
  </button>
</footer>
`;
